### PR TITLE
Smart classroom demo gapi

### PR DESCRIPTION
**Pull request by mpashchenkov**
Demo based on `age_gender_emotion_recognition sample`.

Networks:
- face-detection-retail-0004;
- person-detection-action-recognition-0005.

Networks of OMZ demo (from README):
- [x] face-detection-adas-0001 (face-detection-retail-0004 in YouTube Intel video);
- [ ] landmarks-regression-retail-0009 (WIP);
- [ ] face-reidentification-retail-0095 (WIP, python script);
- [x] person-detection-action-recognition-0005;
- [ ] person-detection-action-recognition-0006 (two-three functions will add to 0005 code);
- [ ] person-detection-raisinghand-recognition-0001 (Didn't touch);
- [ ] person-detection-action-recognition-teacher-0002 (Didn't touch, chinese classroom).

### CODE REVIEW

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        "classroom";    
+    const std::string keys =
+    "{ h help |   | print this help message }"
+    "{ input  |   | Path to an input video file }"
+    "{ fdm    |   | IE face detection model IR }"
+    "{ pam    |   | IE person detection action recognition model IR }" 
+    "{ fdd    |   | IE face detection device }";
+}
+
+namespace custom {
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: Can i move this (kernels, nets) to the appropriate files from main.cpp?

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                                      const cv::GMatDesc &, const cv::GMatDesc &,
+                                      const cv::GMatDesc &, const cv::GMatDesc &) {
+            return cv::empty_array_desc();
+        }
+    };
+    GAPI_OCV_KERNEL(OCVPostProc, PostProc) {
+        static void run(const cv::Mat &in_ssd_result,
+                        const cv::Mat &in_frame,
+                        std::vector<cv::Rect> &out_faces) {
+            const int MAX_PROPOSALS = 200;
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: Constants will be changed to same constants from OMZ demo (Analog of ActionDetectorConfig).
I don't put code from OMZ (it has same result), but if demo should be the same, then i will put it.

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                
+            const cv::Size frame_size = in_frame.size();
+            std::unique_ptr<ActionDetector> detector(new ActionDetector);
+            detector->GetPostProcResult(local_data, det_conf_data, prior_data, action_conf_data, frame_size, out_detections);
+        }
+    };
+}// namespace custom
+    
+int main(int argc, char* argv[]) {
+    cv::CommandLineParser cmd(argc, argv, keys);
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: It should be the same OMZ demo. I got it. I will change it.

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    cv::GComputation pp([]() {
+        cv::GMat in;
+        cv::GMat detections = cv::gapi::infer<custom::Faces>(in);
+        cv::GArray<cv::Rect> faces = custom::PostProc::on(detections, in);
+
+        cv::GMat location;
+        cv::GMat detect_confidences;
+        cv::GMat priorboxes;
+        cv::GMat action_con1;
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: Here. I wanted to use `cv::GArray anchors = {action_con1, action_con2, action_con3, action_con4};`. And put GArray in post processing instead four GMat.
But it didn't work (failed experiment =) ).

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        getBinPath(paPathXML),
+        cmd.get<std::string>("fdd"),
+    }.cfgOutputLayers({"mbox_loc1/out/conv/flat", 
+                       "mbox_main_conf/out/conv/flat/softmax/flat",
+                       "mbox/priorbox",
+                       "out/anchor1",
+                       "out/anchor2",
+                       "out/anchor3",
+                       "out/anchor4"
+                       }); // 0006 have another outputs
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: person-detection-action-recognition-0006 has another outputs (small changes in post processing).

File: demos/smart_classroom_demo_gapi/main.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+    std::string paPathXML = cmd.get<std::string>("pam"); 
+    auto pos_act_net = cv::gapi::ie::Params<custom::PersAction> {
+        paPathXML,
+        getBinPath(paPathXML),
+        cmd.get<std::string>("fdd"),
+    }.cfgOutputLayers({"mbox_loc1/out/conv/flat", 
+                       "mbox_main_conf/out/conv/flat/softmax/flat",
+                       "mbox/priorbox",
+                       "out/anchor1",
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/main.cpp]: Can i get information about dims (n,c,H,W) of output? It is for post processing.

File: demos/smart_classroom_demo_gapi/src/person_action_detector.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+
+#define SSD_LOCATION_RECORD_SIZE 4
+#define SSD_PRIORBOX_RECORD_SIZE 4
+#define NUM_DETECTION_CLASSES 2
+#define POSITIVE_DETECTION_IDX 1
+#define NUM_CANDIDATES 4300
+#define INVALID_TOP_K_IDX -1
+
+void ActionDetector::GetPostProcResult(const float* local_data,
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/src/person_action_detector.cpp:14]: `fetchResults()` from OMZ is our `GAPI_OCV_KERNEL(OCVPersonDetActionRec...)`, 
`GetDetections` from OMZ is our `GetPostProcResult`.

File: demos/smart_classroom_demo_gapi/src/person_action_detector.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    const auto& head_anchors = config_.old_anchors;
+    const int num_heads = head_anchors.size();
+
+    head_ranges_.resize(num_heads + 1);
+    glob_anchor_map_.resize(num_heads);
+    head_step_sizes_.resize(num_heads);
+
+    num_glob_anchors_ = 0;
+    head_ranges_[0] = 0;        
+    glob_anchor_map_[0].resize(head_anchors[0/*head_id*/]);
```
Comment (mpashchenkov) [inline demos/smart_classroom_demo_gapi/src/person_action_detector.cpp:31]: Nulls in `[0]` is basis for 0006.

**Comment by mpashchenkov** [comment]
Result now:

Result

**Comment by mpashchenkov** [comment]
Is the construction of the person-detection-action-recognition correct? 
Can this be taken as a basis?

**Comment by mpashchenkov** [comment]
@dmatveev Please review.

File: demos/smart_classroom_demo_gapi/main.cpp (hunk 1 above)
Comment (dmatveev) [reply demos/smart_classroom_demo_gapi/main.cpp]: Yes you can, but do you actually need this?

**Comment by mpashchenkov** [comment]
New PR in branch with smart_classroom_demo_gapi folder open now.

**Comment by dmatveev** [comment]
If possible, I'd also delete this MR (not just close it)

**Pull request closed without merging**